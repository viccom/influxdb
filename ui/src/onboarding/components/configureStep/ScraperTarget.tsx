// Libraries
import React, {PureComponent, ChangeEvent} from 'react'

// Components
import {
  Form,
  Input,
  Columns,
  Grid,
  ComponentSize,
  Dropdown,
  InputType,
} from 'src/clockface'

interface Props {
  bucket: string
  buckets: string[]
  onSelectBucket: (value: string) => void
  onChangeURL: (value: string) => void
  url: string
}

export class ScraperTarget extends PureComponent<Props> {
  constructor(props: Props) {
    super(props)
  }

  public render() {
    const {bucket, onSelectBucket, url} = this.props

    return (
      <Grid>
        <Grid.Row>
          <Grid.Column widthXS={Columns.Eight} offsetXS={Columns.Two}>
            <Form.Element label="Bucket">
              <Dropdown selectedID={bucket} onChange={onSelectBucket}>
                {this.dropdownBuckets}
              </Dropdown>
            </Form.Element>
          </Grid.Column>
          <Grid.Column widthXS={Columns.Eight} offsetXS={Columns.Two}>
            <Form.Element label="Target URL">
              <Input
                type={InputType.Text}
                value={url}
                onChange={this.handleChangeURL}
                titleText="Target URL"
                size={ComponentSize.Medium}
                autoFocus={true}
              />
            </Form.Element>
          </Grid.Column>
        </Grid.Row>
      </Grid>
    )
  }

  private get dropdownBuckets(): JSX.Element[] {
    const {buckets} = this.props

    return buckets.map(b => (
      <Dropdown.Item key={b} value={b} id={b}>
        {b}
      </Dropdown.Item>
    ))
  }

  private handleChangeURL = (e: ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    this.props.onChangeURL(value)
  }
}

export default ScraperTarget
