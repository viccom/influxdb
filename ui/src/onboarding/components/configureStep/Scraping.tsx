// Libraries
import React, {PureComponent} from 'react'
import {connect} from 'react-redux'

// Components
import {Form} from 'src/clockface'
import FancyScrollbar from 'src/shared/components/fancy_scrollbar/FancyScrollbar'
import OnboardingButtons from 'src/onboarding/components/OnboardingButtons'

// Actions
import {
  setScraperTargetBucket,
  setScraperTargetURL,
  saveScraperTarget,
} from 'src/onboarding/actions/dataLoaders'
import {AppState} from 'src/types/v2/index'
import ScraperTarget from 'src/onboarding/components/configureStep/ScraperTarget'

interface OwnProps {
  onClickNext: () => void
  onClickBack: () => void
  onClickSkip: () => void
}

interface DispatchProps {
  onSetScraperTargetBucket: typeof setScraperTargetBucket
  onSetScraperTargetURL: typeof setScraperTargetURL
  onSaveScraperTarget: typeof saveScraperTarget
}

interface StateProps {
  bucket: string
  url: string
  currentBucket: string
}

type Props = OwnProps & DispatchProps & StateProps

export class Scraping extends PureComponent<Props> {
  public componentDidMount() {
    const {bucket, currentBucket, onSetScraperTargetBucket} = this.props
    if (!bucket) {
      onSetScraperTargetBucket(currentBucket)
    }
  }

  public render() {
    const {
      bucket,
      onClickBack,
      onClickSkip,
      onSetScraperTargetURL,
      url,
    } = this.props

    return (
      <Form onSubmit={this.handleSubmit}>
        <div className="wizard-step--scroll-area">
          <FancyScrollbar autoHide={false}>
            <div className="wizard-step--scroll-content">
              <h3 className="wizard-step--title">Add Scraper Target</h3>
              <h5 className="wizard-step--sub-title">
                Scrapers collect data from multiple targets at regular intervals
                and to write to a bucket
              </h5>
              <ScraperTarget
                bucket={bucket}
                buckets={this.buckets}
                onSelectBucket={this.handleSelectBucket}
                onChangeURL={onSetScraperTargetURL}
                url={url}
              />
            </div>
          </FancyScrollbar>
        </div>
        <OnboardingButtons
          onClickBack={onClickBack}
          onClickSkip={onClickSkip}
          showSkip={true}
          autoFocusNext={false}
          skipButtonText={'Skip'}
        />
      </Form>
    )
  }

  private get buckets(): string[] {
    const {currentBucket} = this.props

    return currentBucket ? [currentBucket] : []
  }

  private handleSelectBucket = (bucket: string) => {
    this.props.onSetScraperTargetBucket(bucket)
  }

  private handleSubmit = async () => {
    await this.props.onSaveScraperTarget()

    this.props.onClickNext()
  }
}

const mstp = ({
  dataLoading: {
    dataLoaders: {scraperTarget},
    steps: {bucket},
  },
}: AppState): StateProps => {
  return {
    currentBucket: bucket,
    bucket: scraperTarget.bucket,
    url: scraperTarget.url,
  }
}

const mdtp: DispatchProps = {
  onSetScraperTargetBucket: setScraperTargetBucket,
  onSetScraperTargetURL: setScraperTargetURL,
  onSaveScraperTarget: saveScraperTarget,
}

export default connect<StateProps, DispatchProps, OwnProps>(
  mstp,
  mdtp
)(Scraping)
