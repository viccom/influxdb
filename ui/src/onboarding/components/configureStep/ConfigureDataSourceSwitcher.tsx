// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import LineProtocol from 'src/onboarding/components/configureStep/lineProtocol/LineProtocol'
import PluginConfigSwitcher from 'src/onboarding/components/configureStep/streaming/PluginConfigSwitcher'
import EmptyDataSourceState from 'src/onboarding/components/configureStep/EmptyDataSourceState'
import Scraping from 'src/onboarding/components/configureStep/Scraping'

// Actions
import {
  updateTelegrafPluginConfig,
  addConfigValue,
  removeConfigValue,
  setConfigArrayValue,
} from 'src/onboarding/actions/dataLoaders'

// Types
import {TelegrafPlugin, DataLoaderType} from 'src/types/v2/dataLoaders'

export interface Props {
  telegrafPlugins: TelegrafPlugin[]
  currentIndex: number
  onUpdateTelegrafPluginConfig: typeof updateTelegrafPluginConfig
  onAddConfigValue: typeof addConfigValue
  onRemoveConfigValue: typeof removeConfigValue
  dataLoaderType: DataLoaderType
  bucket: string
  org: string
  username: string
  onSetConfigArrayValue: typeof setConfigArrayValue
  onClickNext: () => void
  onClickPrevious: () => void
  onClickSkip: () => void
}

@ErrorHandling
class ConfigureDataSourceSwitcher extends PureComponent<Props> {
  public render() {
    const {
      bucket,
      org,
      telegrafPlugins,
      currentIndex,
      dataLoaderType,
      onUpdateTelegrafPluginConfig,
      onAddConfigValue,
      onRemoveConfigValue,
      onSetConfigArrayValue,
      onClickNext,
      onClickPrevious,
      onClickSkip,
    } = this.props

    switch (dataLoaderType) {
      case DataLoaderType.Streaming:
        return (
          <PluginConfigSwitcher
            onUpdateTelegrafPluginConfig={onUpdateTelegrafPluginConfig}
            onRemoveConfigValue={onRemoveConfigValue}
            telegrafPlugins={telegrafPlugins}
            currentIndex={currentIndex}
            onAddConfigValue={onAddConfigValue}
            onSetConfigArrayValue={onSetConfigArrayValue}
            onClickNext={onClickNext}
            onClickPrevious={onClickPrevious}
            onClickSkip={onClickSkip}
          />
        )
      case DataLoaderType.LineProtocol:
        return (
          <LineProtocol
            bucket={bucket}
            org={org}
            onClickNext={onClickNext}
            onClickBack={onClickPrevious}
            onClickSkip={onClickSkip}
          />
        )
      case DataLoaderType.Scraping:
        return (
          <Scraping
            onClickNext={onClickNext}
            onClickBack={onClickPrevious}
            onClickSkip={onClickSkip}
          />
        )
      case DataLoaderType.CSV:
        return <div>{DataLoaderType.CSV}</div>
      default:
        return <EmptyDataSourceState />
    }
  }
}

export default ConfigureDataSourceSwitcher
