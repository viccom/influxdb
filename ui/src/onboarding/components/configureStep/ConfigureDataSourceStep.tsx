// Libraries
import React, {PureComponent} from 'react'
import _ from 'lodash'

// Components
import {ErrorHandling} from 'src/shared/decorators/errors'
import ConfigureDataSourceSwitcher from 'src/onboarding/components/configureStep/ConfigureDataSourceSwitcher'

// Actions
import {setActiveTelegrafPlugin} from 'src/onboarding/actions/dataLoaders'
import {
  updateTelegrafPluginConfig,
  setPluginConfiguration,
  addConfigValue,
  removeConfigValue,
  setConfigArrayValue,
} from 'src/onboarding/actions/dataLoaders'

// Constants
import {StepStatus} from 'src/clockface/constants/wizard'

// Types
import {DataLoaderStepProps} from 'src/dataLoaders/components/DataLoadersWizard'
import {
  TelegrafPlugin,
  DataLoaderType,
  ConfigurationState,
} from 'src/types/v2/dataLoaders'

export interface OwnProps extends DataLoaderStepProps {
  telegrafPlugins: TelegrafPlugin[]
  onSetActiveTelegrafPlugin: typeof setActiveTelegrafPlugin
  onUpdateTelegrafPluginConfig: typeof updateTelegrafPluginConfig
  onSetPluginConfiguration: typeof setPluginConfiguration
  type: DataLoaderType
  onAddConfigValue: typeof addConfigValue
  onRemoveConfigValue: typeof removeConfigValue
  onSetConfigArrayValue: typeof setConfigArrayValue
  bucket: string
  org: string
  username: string
}

type Props = OwnProps

@ErrorHandling
export class ConfigureDataSourceStep extends PureComponent<Props> {
  constructor(props: Props) {
    super(props)
  }

  public render() {
    const {
      telegrafPlugins,
      type,
      substep,
      onUpdateTelegrafPluginConfig,
      onAddConfigValue,
      onRemoveConfigValue,
      onSetConfigArrayValue,
      bucket,
      org,
      username,
    } = this.props

    return (
      <div className="onboarding-step wizard--skippable">
        <ConfigureDataSourceSwitcher
          bucket={bucket}
          org={org}
          username={username}
          telegrafPlugins={telegrafPlugins}
          onUpdateTelegrafPluginConfig={onUpdateTelegrafPluginConfig}
          onAddConfigValue={onAddConfigValue}
          onRemoveConfigValue={onRemoveConfigValue}
          dataLoaderType={type}
          currentIndex={+substep}
          onSetConfigArrayValue={onSetConfigArrayValue}
          onClickNext={this.handleNext}
          onClickPrevious={this.handlePrevious}
          onClickSkip={this.jumpToCompletionStep}
        />
      </div>
    )
  }

  private jumpToCompletionStep = () => {
    const {onSetCurrentStepIndex, stepStatuses, type} = this.props

    this.handleSetStepStatus()

    if (type === DataLoaderType.Streaming) {
      onSetCurrentStepIndex(stepStatuses.length - 2)
    } else {
      onSetCurrentStepIndex(stepStatuses.length - 1)
    }
  }

  private handleNext = async () => {
    const {
      onIncrementCurrentStepIndex,
      onSetActiveTelegrafPlugin,
      onSetPluginConfiguration,
      telegrafPlugins,
      substep,
      currentStepIndex,
      onSetSubstepIndex,
      type,
    } = this.props

    const index = +substep
    const telegrafPlugin = _.get(telegrafPlugins, `${index}.name`)

    onSetPluginConfiguration(telegrafPlugin)
    this.handleSetStepStatus()

    if (type === DataLoaderType.Streaming) {
      if (index >= telegrafPlugins.length - 1) {
        onIncrementCurrentStepIndex()
        onSetActiveTelegrafPlugin('')
      } else {
        const name = _.get(telegrafPlugins, `${index + 1}.name`, '')
        onSetActiveTelegrafPlugin(name)
        onSetSubstepIndex(+currentStepIndex, index + 1)
      }
      return
    }

    onIncrementCurrentStepIndex()
  }

  private handlePrevious = () => {
    const {
      type,
      substep,
      currentStepIndex,
      onSetActiveTelegrafPlugin,
      onSetPluginConfiguration,
      telegrafPlugins,
      onSetSubstepIndex,
      onDecrementCurrentStepIndex,
    } = this.props

    const index = +substep
    const telegrafPlugin = _.get(telegrafPlugins, `${index}.name`)

    if (type === DataLoaderType.Streaming) {
      onSetPluginConfiguration(telegrafPlugin)
      this.handleSetStepStatus()

      if (index > 0) {
        const name = _.get(telegrafPlugins, `${index - 1}.name`)
        onSetActiveTelegrafPlugin(name)
        onSetSubstepIndex(+currentStepIndex, index - 1)
      } else {
        onSetActiveTelegrafPlugin('')
        onSetSubstepIndex(+currentStepIndex - 1, 'streaming')
      }

      return
    }

    onDecrementCurrentStepIndex()
  }

  private handleSetStepStatus = () => {
    const {
      type,
      telegrafPlugins,
      onSetStepStatus,
      currentStepIndex,
    } = this.props

    if (type === DataLoaderType.Streaming) {
      const unconfigured = telegrafPlugins.find(tp => {
        return tp.configured === ConfigurationState.Unconfigured
      })

      if (unconfigured || !telegrafPlugins.length) {
        onSetStepStatus(currentStepIndex, StepStatus.Incomplete)
      } else {
        onSetStepStatus(currentStepIndex, StepStatus.Complete)
      }
    } else {
      onSetStepStatus(currentStepIndex, StepStatus.Complete)
    }
  }
}

export default ConfigureDataSourceStep
