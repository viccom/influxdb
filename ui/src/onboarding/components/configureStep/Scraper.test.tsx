// Libraries
import React from 'react'
import {shallow} from 'enzyme'

// Components
import {ScraperTarget} from 'src/onboarding/components/configureStep/ScraperTarget'

const setup = (override = {}) => {
  const props = {
    bucket: '',
    buckets: [],
    onSelectBucket: jest.fn(),
    onChangeURL: jest.fn(),
    url: '',
    ...override,
  }
  const wrapper = shallow(<ScraperTarget {...props} />)

  return {wrapper}
}

describe('ScraperTarget', () => {
  describe('rendering', () => {
    it('renders!', () => {
      const {wrapper} = setup()
      expect(wrapper.exists()).toBe(true)

      expect(wrapper).toMatchSnapshot()
    })
  })
})
