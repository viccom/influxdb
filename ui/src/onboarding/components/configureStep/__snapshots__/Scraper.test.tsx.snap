// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScraperTarget rendering renders! 1`] = `
<Grid>
  <GridRow>
    <GridColumn
      offsetXS={2}
      widthXS={8}
    >
      <FormElement
        label="Bucket"
      >
        <Dropdown
          buttonColor="default"
          buttonSize="sm"
          maxMenuHeight={250}
          menuColor="sapphire"
          mode="radio"
          onChange={[MockFunction]}
          selectedID=""
          status="default"
          titleText=""
          wrapText={false}
        />
      </FormElement>
    </GridColumn>
    <GridColumn
      offsetXS={2}
      widthXS={8}
    >
      <FormElement
        label="Target URL"
      >
        <Input
          autoFocus={true}
          autocomplete="off"
          disabledTitleText="This input is disabled"
          name=""
          onChange={[Function]}
          placeholder=""
          size="md"
          spellCheck={false}
          status="default"
          titleText="Target URL"
          type="text"
          value=""
        />
      </FormElement>
    </GridColumn>
  </GridRow>
</Grid>
`;
